File: src/server/repositories/field-verification-repository.ts
import { VerificationStatus, UserRole } from "@prisma/client";
import type { Principal } from "../auth";
import { prisma } from "../prisma";

const measurementFields = { id: true, measurementType: true, parameterName: true, standardValue: true, measuredValue: true, unit: true, tolerance: true, calculatedError: true, standardReference: true, passFail: true, recordedAt: true, createdAt: true } as const;
const evidenceFields = { id: true, fileUrl: true, storageKey: true, photoType: true, caption: true, capturedAt: true, isPrimaryEvidence: true, createdAt: true } as const;
const verificationFields = {
  verificationDate: true,
  resultStatus: true,
  findingsSummary: true,
  isCompliant: true,
  recommendedAction: true,
  certificateEligible: true,
  remarks: true,
  createdAt: true,
  updatedAt: true,
  assignment: { select: { status: true, assignedOfficerId: true, applicationId: true, instrumentId: true, application: { select: { applicationNumber: true, status: true, instrumentId: true } }, instrument: { select: { digitalInstrumentId: true, serialNumber: true, instrumentType: true } } } },
  measurements: { select: measurementFields, orderBy: { recordedAt: "asc" } },
  evidencePhotos: { select: evidenceFields, orderBy: { capturedAt: "asc" } },
} as const;

function accessWhere(principal: Principal) {
  if (principal.role === UserRole.ADMINISTRATOR) return {};
  return { verifiedByUserId: principal.userId };
}

export const fieldVerificationRepository = {
  findAccessible(principal: Principal, id: string) {
    return prisma.fieldVerification.findFirst({ where: { id, ...accessWhere(principal) }, select: { ...verificationFields, id: true, assignmentId: true, applicationId: true, instrumentId: true, verifiedByUserId: true } });
  },
  findByAssignment(assignmentId: string) {
    return prisma.fieldVerification.findUnique({ where: { assignmentId }, select: { ...verificationFields, id: true, assignmentId: true, applicationId: true, instrumentId: true, verifiedByUserId: true } });
  },
  create(data: { assignmentId: string; applicationId: string; instrumentId: string; verifiedByUserId: string; verificationDate?: Date; resultStatus: VerificationStatus; findingsSummary?: string; isCompliant?: boolean; recommendedAction?: string; certificateEligible?: boolean; remarks?: string }) {
    return prisma.fieldVerification.create({ data, select: { ...verificationFields, id: true, assignmentId: true, applicationId: true, instrumentId: true, verifiedByUserId: true } });
  },
  update(id: string, data: { resultStatus?: VerificationStatus; findingsSummary?: string | null; isCompliant?: boolean | null; recommendedAction?: string | null; certificateEligible?: boolean | null; remarks?: string | null }) {
    return prisma.fieldVerification.update({ where: { id }, data, select: { ...verificationFields, id: true, assignmentId: true, applicationId: true, instrumentId: true, verifiedByUserId: true } });
  },
  addMeasurement(data: { verificationId: string; instrumentId: string; measurementType: string; parameterName: string; standardValue?: string; measuredValue: string; unit?: string; tolerance?: string; calculatedError?: string; standardReference?: string; passFail: boolean }) {
    return prisma.measurementRecord.create({ data, select: measurementFields });
  },
  listMeasurements(verificationId: string) {
    return prisma.measurementRecord.findMany({ where: { verificationId }, select: measurementFields, orderBy: { recordedAt: "asc" } });
  },
  updateMeasurement(id: string, data: { measurementType?: string; parameterName?: string; standardValue?: string | null; measuredValue?: string; unit?: string | null; tolerance?: string | null; calculatedError?: string | null; standardReference?: string | null; passFail?: boolean }) {
    return prisma.measurementRecord.update({ where: { id }, data, select: measurementFields });
  },
  findMeasurement(id: string) {
    return prisma.measurementRecord.findUnique({ where: { id }, select: { ...measurementFields, verification: { select: { id: true, verifiedByUserId: true, assignmentId: true } } } });
  },
  deleteMeasurement(id: string) { return prisma.measurementRecord.delete({ where: { id } }); },
  addEvidence(data: { verificationId: string; instrumentId: string; fileUrl: string; storageKey?: string; photoType: string; caption?: string; capturedAt?: Date; isPrimaryEvidence?: boolean }) {
    return prisma.evidencePhoto.create({ data, select: evidenceFields });
  },
  listEvidence(verificationId: string) {
    return prisma.evidencePhoto.findMany({ where: { verificationId }, select: evidenceFields, orderBy: { capturedAt: "asc" } });
  },
  findEvidence(id: string) {
    return prisma.evidencePhoto.findUnique({ where: { id }, select: { ...evidenceFields, verification: { select: { id: true, verifiedByUserId: true, assignmentId: true } } } });
  },
  deleteEvidence(id: string) { return prisma.evidencePhoto.delete({ where: { id } }); },
};

File: mobile_app/lib/features/field_verification/field_verification_repository.dart
import '../../core/network/api_client.dart';
import 'dart:io';

class FieldVerificationRepository {
  final ApiClient _apiClient;

  FieldVerificationRepository(this._apiClient);

  Future<FieldVerificationModel> getOrCreateVerification(String assignmentId) async {
    try {
      return await getVerificationByAssignment(assignmentId);
    } catch (e) {
      return await createVerification(assignmentId);
    }
  }

  Future<FieldVerificationModel> getVerificationByAssignment(String assignmentId) async {
    final response = await _apiClient.request<Map<String, dynamic>>(
      path: '/assignments/$assignmentId/field-verification',
      method: 'GET',
    );
    return FieldVerificationModel.fromJson(response);
  }

  Future<FieldVerificationModel> createVerification(String assignmentId) async {
    final response = await _apiClient.request<Map<String, dynamic>>(
      path: '/assignments/$assignmentId/field-verification',
      method: 'POST',
      body: {
        'verificationDate': DateTime.now().toIso8601String(),
        'findingsSummary': '',
        'remarks': '',
      },
    );
    return FieldVerificationModel.fromJson(response);
  }

  Future<FieldVerificationModel> getVerification(String verificationId) async {
    final response = await _apiClient.request<Map<String, dynamic>>(
      path: '/field-verifications/$verificationId',
      method: 'GET',
    );
    return FieldVerificationModel.fromJson(response);
  }

  Future<MeasurementModel> addMeasurement(String verificationId, Map<String, dynamic> data) async {
    final response = await _apiClient.request<Map<String, dynamic>>(
      path: '/field-verifications/$verificationId/measurements',
      method: 'POST',
      body: data,
    );
    return MeasurementModel.fromJson(response);
  }

  Future<void> submitVerification(String verificationId, Map<String, dynamic> data) async {
    await _apiClient.request(
      path: '/field-verifications/$verificationId/submit',
      method: 'POST',
      body: data,
    );
  }

  Future<List<MeasurementModel>> listMeasurements(String verificationId) async {
    final response = await _apiClient.request<List<dynamic>>(
      path: '/field-verifications/$verificationId/measurements',
      method: 'GET',
    );
    return response.map((json) => MeasurementModel.fromJson(json)).toList();
  }

  Future<EvidencePhotoModel> addEvidence(String verificationId, {
    required String filePath,
    required String category,
    String? caption,
    bool isPrimary = false,
  }) async {
    final response = await _apiClient.multipartRequest<Map<String, dynamic>>(
      path: '/field-verifications/$verificationId/evidence',
      fields: {
        'photoType': category,
        'caption': caption ?? '',
        'isPrimaryEvidence': isPrimary ? 'true' : 'false',
      },
      files: {
        'file': File(filePath),
      },
    );
    return EvidencePhotoModel.fromJson(response);
  }
}

class FieldVerificationModel {
  final String id;
  final String status;
  final String verificationDate;
  final String? resultStatus;
  final String? findingsSummary;
  final bool? isCompliant;
  final bool? certificateEligible;
  final String? remarks;
  final String assignmentId;
  final List<EvidencePhotoModel> evidencePhotos;

  FieldVerificationModel({
    required this.id,
    required this.status,
    required this.verificationDate,
    this.resultStatus,
    this.findingsSummary,
    this.isCompliant,
    this.certificateEligible,
    this.remarks,
    required this.assignmentId,
    this.evidencePhotos = const [],
  });

  factory FieldVerificationModel.fromJson(Map<String, dynamic> json) {
    return FieldVerificationModel(
      id: json['id'],
      status: json['status'],
      verificationDate: json['verificationDate'],
      resultStatus: json['resultStatus'],
      findingsSummary: json['findingsSummary'],
      isCompliant: json['isCompliant'],
      certificateEligible: json['certificateEligible'],
      remarks: json['remarks'],
      assignmentId: json['assignmentId'],
      evidencePhotos: (json['evidencePhotos'] as List? ?? [])
          .map((e) => EvidencePhotoModel.fromJson(e))
          .toList(),
    );
  }
}

class EvidencePhotoModel {
  final String id;
  final String fileUrl;
  final String photoType;
  final String? caption;
  final String capturedAt;
  final bool isPrimaryEvidence;

  EvidencePhotoModel({
    required this.id,
    required this.fileUrl,
    required this.photoType,
    this.caption,
    required this.capturedAt,
    required this.isPrimaryEvidence,
  });

  factory EvidencePhotoModel.fromJson(Map<String, dynamic> json) {
    return EvidencePhotoModel(
      id: json['id'],
      fileUrl: json['fileUrl'],
      photoType: json['photoType'],
      caption: json['caption'],
      capturedAt: json['capturedAt'],
      isPrimaryEvidence: json['isPrimaryEvidence'] ?? false,
    );
  }
}

class MeasurementModel {
  final String id;
  final String measurementType;
  final String parameterName;
  final String? standardValue;
  final String measuredValue;
  final String? unit;
  final String? tolerance;
  final String? standardReference;
  final String? calculatedError;
  final bool passFail;
  final String recordedAt;

  MeasurementModel({
    required this.id,
    required this.measurementType,
    required this.parameterName,
    this.standardValue,
    required this.measuredValue,
    this.unit,
    this.tolerance,
    this.standardReference,
    this.calculatedError,
    required this.passFail,
    required this.recordedAt,
  });

  factory MeasurementModel.fromJson(Map<String, dynamic> json) {
    return MeasurementModel(
      id: json['id'],
      measurementType: json['measurementType'],
      parameterName: json['parameterName'],
      standardValue: json['standardValue'],
      measuredValue: json['measuredValue'],
      unit: json['unit'],
      tolerance: json['tolerance'],
      standardReference: json['standardReference'],
      calculatedError: json['calculatedError'],
      passFail: json['passFail'] ?? false,
      recordedAt: json['recordedAt'],
    );
  }
}
